integration-tests/src/tests/client/features/multinode_stateless_validators.rs

```diff
@@ -0,0 +1,674 @@
+        let shutting_down = test.shutting_down();
+        test.sender().for_index(idx).send_adhoc_event("start_shards_manager", move |data| {
+            data.shards_manager.periodically_resend_chunk_requests(
+                &mut sender.into_delayed_action_runner(shutting_down),
+            );
+        });
+
+        test.sender().for_index(idx).send_adhoc_event("start_state_sync_dumper", move |data| {
+            data.state_sync_dumper.start().unwrap();
+        });
+    }
+
+    // Give it some condition to stop running at. Here we run the test until the first client
+    // reaches height 10003, with a timeout of 5sec (failing if it doesn't reach 10003 in time).
+    test.run_until(
+        |data| data[0].client.client.chain.head().unwrap().height == 10003,
+        Duration::seconds(5),
+    );
+    for idx in 0..NUM_VALIDATORS {
+        test.sender().for_index(idx).send_adhoc_event("assertions", |data| {
+            let chain = &data.client.client.chain;
+            let block = chain.get_block_by_height(10002).unwrap();
+            assert_eq!(
+                block.header().chunk_mask(),
+                &(0..NUM_SHARDS).map(|_| true).collect::<Vec<_>>()
+            );
+        })
+    }
+    test.run_instant();
+
+    // Capture the initial validator info in the first epoch.
+    let initial_epoch_id = test.data[0].client.client.chain.head().unwrap().epoch_id;
+
+    let mut balances = accounts
+        .iter()
+        .cloned()
+        .map(|account| (account, initial_balance))
+        .collect::<HashMap<_, _>>();
+
+    let anchor_hash = *test.data[0].client.client.chain.get_block_by_height(10002).unwrap().hash();
+
+    // Run send-money transactions between "non-validator" accounts.
+    for i in NUM_VALIDATORS..NUM_ACCOUNTS {
+        let amount = ONE_NEAR * (i as u128 + 1);
+        let tx = SignedTransaction::send_money(
+            1,
+            accounts[i].clone(),
+            accounts[(i + 1) % NUM_ACCOUNTS].clone(),
+            &create_user_test_signer(&accounts[i]),
+            amount,
+            anchor_hash,
+        );
+        *balances.get_mut(&accounts[i]).unwrap() -= amount;
+        *balances.get_mut(&accounts[(i + 1) % NUM_ACCOUNTS]).unwrap() += amount;
+        drop(
+            test.sender()
+                .for_index(i % NUM_VALIDATORS)
+                .with_additional_delay(Duration::milliseconds(300 * i as i64))
+                .into_wrapped_multi_sender::<ClientSenderForNetworkMessage, ClientSenderForNetwork>(
+                )
+                .send_async(ProcessTxRequest {
+                    transaction: tx,
+                    is_forwarded: false,
+                    check_only: false,
+                }),
+        );
+    }
+
+    // Run the chain some time until the following happen both:
+    // 1. All transactions are processed (so we run for some duration).
+    // 2. We transition to a different epoch then prev_epoch_id. (so we run until the epoch id changes)
```

Comment:
Oh. I thought that `test.run_for(Duration::seconds(20));` is the only assertion until I saw another one below. Should we move this comment above that assertion?
Also looks like we should assert that all transactions were actually processed, like, count number of final transaction outcomes somewhere.

Reply:
Updated the comments, but transaction outcome checking is currently not working due to single shard tracking (so the code currently checks the final balances to see if all the transactions took effect).